bodhi/server/consumers/pungi_wrapper.py

```diff
@@ -0,0 +1,383 @@
+                raise Exception(ex)
+        try:
+            os.symlink(self.compose_dir, symlink)
+        except Exception as ex:
+            self.compose.log_error("Couldn't create latest symlink: %s" % ex)
+            raise
+
+
+class VariantsConfig(object):
+
+    """
+    This class generates a variants config which can be used with the compose
+    object to generate a dnf repo. Right now it only works with modules.
+    """
+
+    def __init__(self, updates, builds, variant_id="Server",
+                 arches=['x86_64', 'armhfp', 'aarch64', 'i386', 'ppc64', 'ppc64le', 's390x']):
+        """
+        Args:
+              updates (list) - list of Update objects
+              builds (list) - list of ModularBuild objects
+              variant_id (str) - the type of variant which will be composed
+              arches (list) - architectures which should be included in compose
+        """
+        self.updates = updates
+        self.builds = builds
+        self.modules = self._generate_module_list()
+        self.arches = arches
+        self.headers = [
+            # https://pagure.io/pungi/blob/master/f/share/variants.dtd
+            '<?xml version="1.0" encoding="UTF-8"?>',
+            ('<!DOCTYPE variants PUBLIC "-//Red Hat, Inc.//DTD '
+             'Variants info//EN" "variants2012.dtd">'),
+        ]
+        self.body = [
+            '<variants>',
+            '<variant id="%s" name="%s" type="variant">' % (variant_id, variant_id),
+            '<arches>',
+            ''.join(['<arch>%s</arch>' % arch for arch in self.arches]),
+            '</arches>',
+            '<modules>',
+            ''.join(['<module>%s</module>' % module for module in self.modules]),
+            '</modules>',
+            '</variant>',
+            '</variants>'
+        ]
+
+    def _generate_module_list(self):
+        """
+        Generates a list of NSV which should be used for pungi modular compose
+
+        Returns:
+          list: list of NSV string which should be composed
+        """
+        newest_builds = {}
+        # we loop through builds so we get rid of older builds and get only
+        # a dict with the newest builds
+        for build in self.builds:
+            nsv = build.nvr.rsplit('-', 1)
+            ns = nsv[0]
+            version = nsv[1]
+
+            if ns in newest_builds:
+                curr_version = newest_builds[ns]
+                if int(curr_version) < int(version):
+                    newest_builds[ns] = version
+            else:
+                newest_builds[ns] = version
+
+        # make sure that the modules we want to update get their correct versions
+        for update in self.updates:
+            for build in update.builds:
+                nsv = build.nvr.rsplit('-', 1)
+                ns = nsv[0]
+                version = nsv[1]
+                newest_builds[ns] = version
+
+        module_list = ["%s-%s" % (nstream, v) for nstream, v in newest_builds.iteritems()]
```

Comment:
Will there be problems for the client if the list of modules here isn't the same as the list of modules on release day? i.e., if there are modules present in one that are not in the other, which would happen if a package were added via update that wasn't there on release day?

Reply:
That had better be ok.  If not, it's a dnf issue.  I fully expect the whole stack to support introduction of new modules post-GA.